yksgi/src/lib.rs

```diff
@@ -149,6 +168,271 @@ impl SGInterp {
+            // pushed RBP above (which includes adjusting RBP) we need to temporarily re-adjust our
+            // pointer. This is because the CSR index calculates from the bottom of the frame, not
+            // from RBP. For example, a typical prologue looks like this:
+            //   push rbp
+            //   mov rbp, rsp
+            //   push rbx     # this has index -2
+            //   push r14     # this has index -3
+            if i > 0 {
+                for (reg, idx) in &pinfo.csrs {
+                    let mut tmp = unsafe { rbp.offset(isize::try_from(idx * 8).unwrap()) };
+                    if pinfo.hasfp {
+                        tmp = unsafe { tmp.offset(8) };
+                    }
+                    let val = registers[usize::from(*reg)];
+                    unsafe { ptr::write(tmp as *mut u64, val) };
+                }
+            }
+
+            // WRITE STACKMAP LOCATIONS.
+            // Now write all live variables to the new stack in the order they are listed in the
+            // AOT stackmap call.
+            let smcall = unsafe { get_stackmap_call(frame.pc) };
+            for (j, lv) in rec.live_vars.iter().enumerate() {
+                // Adjust the operand index by 2 to skip stackmap ID and shadow bytes.
+                let op = smcall.get_operand(u32::try_from(j+2).unwrap());
+                let val = if let Some(val) = frame.get(&op) {
+                    val.val
+                } else {
+                    todo!()
+                    //println!("probably return pointer which will disappear");
+                    //continue;
+                };
+                let l = if lv.len() == 1 {
+                    lv.get(0).unwrap()
+                } else {
+                    todo!("Deal with multi register locations");
+                };
+
+                // Iterate over all locations. Register locations just update the current value in
+                // the registers vector. Direct locations make up part of the stack so need to
+                // written to the allocated memory. Other locations we haven't encountered yet, so
+                // will deal with them as they appear.
+                match l {
+                    SMLocation::Register(reg, _size) => {
+                        registers[usize::from(*reg)] = val;
+                    }
+                    SMLocation::Direct(reg, off, _) => {
+                        if i == 0 {
+                            // skip first frame
+                            continue;
+                        }
+                        debug_assert!(op.is_alloca());
+
+                        // The sizes reported by the stackmap aren't always correct. But we can get
+                        // the correct size from the IR.
+                        let eltype = unsafe { LLVMGetAllocatedType(op.get()) };
+                        let size = i64::try_from(unsafe { LLVMABISizeOfType(layout, eltype)}).unwrap();
```

Comment:
Do we need a signed size here?

Reply:
Can allocas have negative sizes?

Reply:
Oh, I see. You are saying we don't need an `i64` here. And I agree.

Reply:
What type does `LLVMABISizeOfType` return? I wonder if the `try_from` is even needed now.

Reply:
Good spot. It's `c_ulonglong` which means we don't need the `try_from`.

Fixed in [cfe5de7](https://github.com/ykjit/yk/pull/578/commits/cfe5de756bc0bd05c3f44dc7a0298e3a5ad1233a)